### PR TITLE
AWS SAM Template for Cache Extension demo

### PR DESCRIPTION
**Pull request by vamsivikash**
*Issue #, if available:*
#28 

*Description of changes:*

Added SAM Template for deploying the sample Lambda application, Dynamo DB table and the Go Executable as a Lambda layer.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cache-extension-demo/SAM/template.yaml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    DependsOn: 
+      - DynamoDBTable
+      - LambdaRole
+    Properties:
+      Handler: index.handler
+      Role: 
+        Fn::GetAtt: [ LambdaRole , "Arn" ]
+      Runtime: nodejs12.x
+      Timeout: 15
+      Code: LambdaCode/
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml]: Can we rename this directory into something a bit more clear of populating the table with sample data?

File: cache-extension-demo/SAM/package.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/package.json]: What's the need for this package.json?

File: cache-extension-demo/SAM/LambdaCode/package.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "lambdacode",
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/LambdaCode/package.json]: Can we rename to something more meaningful?

Comment (vamsivikash) [reply cache-extension-demo/SAM/LambdaCode/package.json]: Modified it to "database-entry"

File: cache-extension-demo/SAM/package.json (hunk 2 above)
Comment (vamsivikash) [reply cache-extension-demo/SAM/package.json]: Removed this file

File: cache-extension-demo/example-function/package.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "database-entry",
```
Comment (santiagocardenas) [inline cache-extension-demo/example-function/package.json]: Is this file needed? If so, can it be updated to match its actual name and purpose?

File: cache-extension-demo/README.md [hunk 5]
```diff
@@ -74,16 +107,6 @@ aws dynamodb put-item \
 ```
 > Note: Dynamodb values are stored under the key generated using the format `<table><hyphen><hashkey><hyphen><rangekey>`
 
-## Compile package and dependencies
```
Comment (santiagocardenas) [inline cache-extension-demo/README.md:77]: Is this really no longer necessary? How would someone get the compiled Go binary?

File: cache-extension-demo/SAM/template.yaml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      Runtime: nodejs12.x
+      Environment: 
+        Variables:
+          databaseName: 
+            Ref: DatabaseName
+      Layers: 
+        - !Ref CacheExtensionLayer
+      Policies:
+        - AWSLambdaExecute
+        - Version: '2012-10-17'
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml]: Can we use SAM managed policy here?

File: cache-extension-demo/SAM/template.yaml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            Service:
+            - lambda.amazonaws.com
+          Action:
+          - sts:AssumeRole
+      Path: "/"
+      ManagedPolicyArns: 
+        - "arn:aws:iam::aws:policy/AWSLambdaExecute"
+      Policies: 
+        - PolicyName: dynamodbAccessRole
+          PolicyDocument:
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml:95]: Can we use SAM managed policy here?

File: cache-extension-demo/SAM/template.yaml [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+              - dynamodb:PutItem
+              Resource: 
+                !Join
+                  - ''
+                  - - !Sub 'arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/'
+                    - Ref: DatabaseName
+
+  # PutEntry Lambda function to put data to Dynamo DB Table 
+  DatabaseEntry:
+    Type: AWS::Lambda::Function
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml]: Can this be a SAM function?

File: cache-extension-demo/SAM/template.yaml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        - 
+          AttributeName: "pKey"
+          KeyType: "HASH"
+        - 
+          AttributeName: "sKey"
+          KeyType: "RANGE"
+
+
+  # Lambda IAM Role to put data to DynamoDB 
+  LambdaRole:
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml]: Can this be a SAM managed policy in the function?

File: cache-extension-demo/SAM/template.yaml (hunk 8 above)
Comment (vamsivikash) [reply cache-extension-demo/SAM/template.yaml]: This is a custom resource to put the data to the DDB, so it's not possible as SAM unfortunately

File: cache-extension-demo/SAM/template.yaml (hunk 9 above)
Comment (vamsivikash) [reply cache-extension-demo/SAM/template.yaml]: SAM Managed policy has additional actions like "dynamodb:UpdateItem", "dynamodb:BatchWriteItem" which are not needed for this demo. Restricted it just to putItem

File: cache-extension-demo/README.md (hunk 5 above)
Comment (santiagocardenas) [reply cache-extension-demo/README.md:77]: I see that it was moved.

File: cache-extension-demo/example-function/package.json (hunk 4 above)
Comment (vamsivikash) [reply cache-extension-demo/example-function/package.json]: Updated to match the correct description and name

File: cache-extension-demo/README.md (hunk 5 above)
Comment (vamsivikash) [reply cache-extension-demo/README.md:77]: Moved it to the top section before Deployment steps

File: cache-extension-demo/SAM/template.yaml [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+Resources:
+  # Sample Dynamo DB Table 
+  DynamoDBTable: 
+    Type: AWS::DynamoDB::Table
+    Properties:
+      TableName: 
+        Ref: DatabaseName
+      BillingMode: PAY_PER_REQUEST
+      AttributeDefinitions:
+        -
```
Comment (santiagocardenas) [inline cache-extension-demo/SAM/template.yaml]: Move up 1 line for consistency
